ICSharpCode.Decompiler/CSharp/ExpressionBuilder.cs

```diff
@@ -1249,6 +1249,21 @@ protected internal override TranslatedExpression VisitBinaryNumericInstruction(B
 			{
 				return null;
 			}
+			if (context.TypeHint.Kind == TypeKind.Pointer)
+			{
+				// We use the type hint if one of the following is true:
+				// * The current element type is a non-primitive struct.
+				// * The current element type has a different size than the type hint element type.
+				// This prevents the type hint from overriding in undesirable situations (eg changing char* to short*).
+
+				var typeHint = (PointerType)context.TypeHint;
+				int elementTypeSize = pointerType.ElementType.GetSize();
+				if ((elementTypeSize == 0 || typeHint.ElementType.GetSize() != elementTypeSize)
+					&& GetPointerArithmeticOffset(byteOffsetInst, byteOffsetExpr, typeHint.ElementType, inst.CheckForOverflow) != null)
```

Comment:
The repeated call to `GetPointerArithmeticOffset` is a bit unfortunate, as it includes the recursive call to `Translate` for the offset instruction -> potential for exponential runtime (in the unlikely case that the offsetExpr contains further pointer arithmetic in a nested expression).

I think a call to `PointerArithmeticOffset.Detect` would be sufficient here, as `GetPointerArithmeticOffset` returns null iff the detection returned null.

Alternatively you could use the `TranslatedExpression` returned here without calling `GetPointerArithmeticOffset` again.